mlflow/langchain/langchain_tracer.py

```diff
@@ -162,6 +163,24 @@ def _end_span(
                 status=status,
             )
 
+    def _is_trace_active(self, request_id: str) -> bool:
+        """Check if a trace with the given request ID is active (i.e. not ended yet)"""
+        return (
+            # Case 1: The root span is started by this callback, the ID
+            # should be in the active list, otherwise it's already ended.
+            request_id in self._active_request_ids
+            # Case 2: The root span is created by fluent API outside this callback.
+            # In this case, we check the context to see if the trace is active or not.
+            or (
+                (active_span := mlflow.get_current_active_span())
+                and (active_span.request_id == request_id)
+            )
+            # Case 3: The root span is created by client API outside this callback.
```

Comment:
The case 3 logic (root span initiated external to the callback handler) wherein the trace is ended prior to running the callback will simply create a new trace silently from the callback context without notification, right? 

If I do this the following I should expect that I get 2 traces that are completely independent of one another. 

```python
import mlflow

from langchain_core.prompts import ChatPromptTemplate
from langchain_openai import ChatOpenAI

mlflow.langchain.autolog()

llm = ChatOpenAI(temperature=0.9)
prompt = ChatPromptTemplate.from_messages(
            [
                ("system", "You are a helpful assistant answering math questions."),
                ("placeholder", "{chat_history}"),
                ("human", "{input}"),
                ("placeholder", "{agent_scratchpad}"),
            ]
        )
chain = prompt | llm 

trace = client.start_trace(name="Client Initiatilized Trace")

@mlflow.trace(name="Fluent")
def run(message):
    span = client.start_span(request_id=trace.request_id, name="Chat Call", parent_id=trace.span_id, attributes={"a": "a"}, inputs=message)
    invocation = chain.invoke(message)
    client.end_span(span_id=span.span_id, request_id=trace.request_id, outputs=invocation)
    client.end_trace(request_id=trace.request_id)
    
    return invocation

run("How far can a baseball be thrown by a mammal?")
```
For trace client uses, it should be irrelevant if the trace is active or not, right? It's the fluent API that we need to address the proper closure. 

Was there a different usage of the Client API within the scope of the callback that you were referring to here?

Reply:
@BenWilson2 

> If I do this the following I should expect that I get 2 traces that are completely independent of one another.

The case 3 I referred here is a slightly different pattern. It is where the user passes `parent_span` to the `MlfowLangChainTracer` constructor (not autolog).

```
span = client.start_trace(request_id=trace.request_id, name="parent")
invocation = chain.invoke(message, config={"callbacks": [MlflowLangChainTracer(parent_span=span)]
client.end_trace(request_id=span.request_id)
```
In this case, the LangChain spans are generated under the root span the user creates.

For your case (where client is used with autolog), yes there should be two different traces🙂